### PR TITLE
feat(vlm): add Gemma 4 image support

### PR DESCRIPTION
**Pull request by Yanko-7**
## Motivation

`Gemma4ForConditionalGeneration` currently uses the text-only implementation. Enable Gemma 4 image inputs through the existing in-model multimodal pipeline.

## Modifications

- Add the vision configuration, encoder, checkpoint weight mappings, and image processor adapter.
- Register Gemma 4 as a multimodal model and reuse shared vision attention, sharding, and lane packing.
- Align overflow buckets to the 3×3 pooling unit so packed images preserve complete pooling groups.
- Support image inputs and explicitly reject unsupported audio/video inputs.

### Test Environment

| Field | Value |
|---|---|
| Hardware | TPU v7x-8, single host, 4 chips / 8 JAX devices |
| Model | `google/gemma-4-31B-it` (revision `842da3794eaa0b77d5f08bae87a17459d91ff475`) |
| Parallelism | DP4 × effective TP2 (`--tp-size 8 --dp-size 4`) |
| Precision | BF16 model, vision tower, and KV cache; no quantization |
| Context / max sequence | 32,768 / 32,768 |
| Vision pipeline | In-model ViT (`--vision-encoder-parallel dp`), PIL image decoding, 2 processor workers |
| Evaluator | `evalscope[app,perf]==1.5.1`, OpenAI-compatible API |
| Dataset source | Hugging Face `MMMU/MMMU` (validation split, 900 questions across 30 subjects) |
| Generation | `max_tokens=8192`, `temperature=0.0`, `stream=False`, `seed=42` |
| Evaluation batch size | 128 |

### Deployment Command

```bash
python -u -m sgl_jax.launch_server \
  --model-path google/gemma-4-31B-it \
  --served-model-name google/gemma-4-31B-it \
  --trust-remote-code \
  --skip-server-warmup \
  --device tpu \
  --tp-size 8 --dp-size 4 \
  --dtype bfloat16 --kv-cache-dtype bf16 \
  --context-length 32768 --max-seq-len 32768 \
  --max-running-requests 256 \
  --max-prefill-tokens 16384 --chunked-prefill-size 4096 \
  --mem-fraction-static 0.8 --page-size 128 \
  --disable-radix-cache \
  --vision-encoder-parallel dp \
  --mm-processor-worker-num 2 \
  --random-seed 0 \
  --host 0.0.0.0 --port 30000
```

### Benchmark Command

```python
from evalscope import TaskConfig, run_task

task_cfg = TaskConfig(
    model="google/gemma-4-31B-it",
    api_url="http://127.0.0.1:30000/v1",
    api_key="EMPTY",
    eval_type="openai_api",
    datasets=["mmmu"],
    dataset_hub="huggingface",
    dataset_args={
        "mmmu": {
            "dataset_id": "MMMU/MMMU",
            "eval_split": "validation",
        }
    },
    eval_batch_size=128,
    generation_config={
        "max_tokens": 8192,
        "temperature": 0.0,
        "stream": False,
    },
    seed=42,
    work_dir="./outputs/gemma4_31b_mmmu",
)

run_task(task_cfg=task_cfg)
```

### Test Results

| Dataset | Metric | Samples | Score |
|---|---|---:|---:|
| MMMU Val | Mean Accuracy | 900 | **77.89%** |

*All 900 validation questions were evaluated with zero request/generation errors (`num_predictions: 900`, `unique_ids: 900`, `errors: []`).*

<details>
<summary><b>Breakdown by Discipline & Subject</b></summary>

#### Discipline Summary

| Discipline | Samples | Accuracy |
|---|---:|---:|
| Art & Design | 120 | 72.50% |
| Business | 150 | 80.00% |
| Science | 150 | 82.67% |
| Health & Medicine | 150 | 80.00% |
| Humanities & Social Science | 120 | 84.17% |
| Tech & Engineering | 210 | 70.95% |
| **OVERALL** | **900** | **77.89%** (701/900) |

#### Subject Highlights

- **100%**: Physics (30/30), Public Health (30/30)
- **90%+**: Art Theory (28/30), Economics (28/30), Marketing (28/30), Design (27/30), Pharmacy (27/30)
- **80%+**: Biology (26/30), History (26/30), Literature (26/30), Accounting (25/30), Computer Science (25/30), Energy & Power (25/30), Math (25/30), Sociology (25/30), Psychology (24/30)

</details>

## Benchmarking

- TPU v7x-8 (4 chips / 8 devices), TP8/DP4, BF16 weights/KV, context 2048; server max-running 256, memory fraction 0.8, prefill 16384/chunk 4096, page 128, radix disabled, vision DP, 2 image workers, PIL decoding.
- Client: `sglang-oai-chat`, `--max-concurrency 100 --request-rate inf`, 1,000 requests per run, 500 output tokens/request, seed 0. One warmup and cache flush per run; image then text on the same server.
- Image: 1 random 512×512 JPEG + random-input-len 1024. Text: random-input-len 1465, approximately matching image total input tokens.

| Metric | Pure text | Image + text |
|---|---:|---:|
| Successful requests | 1,000/1,000 | 1,000/1,000 |
| Mean input tokens | 1,465.000 | 1,465.293 |
| Total output tokens | 500,000 | 500,000 |
| Duration (s) | 161.814 | 168.415 |
| Request throughput (req/s) | 6.180 | 5.938 |
| Output throughput (tok/s) | 3,089.975 | 2,968.849 |
| Mean / P99 TTFT (ms) | 2,754.475 / 4,788.329 | 3,175.533 / 5,459.032 |
| Mean / P99 TPOT (ms) | 26.880 / 31.032 | 27.129 / 32.854 |
| Mean / P99 E2E (ms) | 16,167.463 / 16,224.250 | 16,712.972 / 17,308.431 |

No recorded client errors. Image + text throughput was 3.92% lower and mean TTFT 421.1 ms higher. Single trial per modality; startup excluded; no accuracy claim.

Environment: JAX 0.11.1, libtpu 0.0.46.1, Flax 0.12.9, torch 2.14.0+cpu. Full multimodal/TPU extras installed, then transformers 5.5.0 and huggingface-hub 1.32.0 used for Gemma4 support. These exceed repository pins; `pip check` also reports decord's unsupported platform. Image preflight and both benchmarks passed; audio/video untested.

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

**Comment by JamesBrianD** [comment]
The benchmark uses Transformers 5.5.0, but python/pyproject.toml still pins transformers~=4.57.1, which does not provide the HF Gemma4Processor. Registering this adapter now routes Gemma 4 through AutoProcessor.from_pretrained(), including text-only serving.

Could you update the dependency constraints to a compatible version, including huggingface-hub as needed, and verify both text-only startup and an image request in a clean environment installed from this PR?

**Review by JamesBrianD — APPROVED** [review]
